feat: Add support for regular before, after and error hooks

marshallswain opened the pull request:
@daffl let's debug these error hooks this week.

Commit:
`collect` hooks util. Not hitting error hooks

daffl committed:
Fix error hook handling and update tests

daffl commented:
Alright, got it sorted, the error hooks should've gone first. We still need to add docs.

marshallswain commented:
Awesome.  Thanks for fixing that.  I'll work on docs.

marshallswain committed:
docs: complete refactor

marshallswain committed:
fix: cleaner utilities for regular hooks & collect

This pulls over the code for Regular hooks from this PR: https://github.com/feathersjs/feathers/pull/2483

marshallswain committed:
Merge branch 'master' into collect-hooks
